sensor/common/virtualmachine/service_impl.go

```diff
@@ -0,0 +1,148 @@
+		fromDataSource: make(chan *storage.VirtualMachine, 10),
+	}
+}
+
+type serviceImpl struct {
+	sensor.UnimplementedVirtualMachineServiceServer
+	toCentral      chan *message.ExpiringMessage
+	stopper        concurrency.Stopper
+	fromDataSource chan *storage.VirtualMachine
+}
+
+func (s *serviceImpl) Name() string {
+	return "virtualMachine.serviceImpl"
+}
+
+// RegisterServiceServer registers this service with the given gRPC Server.
+func (s *serviceImpl) RegisterServiceServer(grpcServer *grpc.Server) {
+	sensor.RegisterVirtualMachineServiceServer(grpcServer, s)
+}
+
+// RegisterServiceHandler implements the APIService interface, but the agent does not accept calls over the gRPC gateway
+func (s *serviceImpl) RegisterServiceHandler(context.Context, *runtime.ServeMux, *grpc.ClientConn) error {
+	return nil
+}
+
+// AuthFuncOverride specifies the auth criteria for this API.
+func (s *serviceImpl) AuthFuncOverride(ctx context.Context, fullMethodName string) (context.Context, error) {
+	if err := idcheck.CollectorOnly().Authorized(ctx, fullMethodName); err != nil {
+		return ctx, errors.Wrapf(err, "virtual machine authorization for %q", fullMethodName)
+	}
+	return ctx, nil
+}
+
+func (s *serviceImpl) UpsertVirtualMachine(ctx context.Context, req *sensor.UpsertVirtualMachineRequest) (*sensor.UpsertVirtualMachineResponse, error) {
+	if s.toCentral == nil {
+		return &sensor.UpsertVirtualMachineResponse{
+			Success: false,
+		}, errors.New("Connection to Central is not ready")
+	}
+
+	if req.VirtualMachine == nil {
+		return &sensor.UpsertVirtualMachineResponse{
+			Success: false,
+		}, errors.New("VirtualMachine in request cannot be nil")
+	}
+
+	log.Debugf("Upserting virtual machine: %s", req.VirtualMachine.Id)
+	s.fromDataSource <- req.VirtualMachine
+	return &sensor.UpsertVirtualMachineResponse{
+		Success: true,
+	}, nil
+}
+
+func (s *serviceImpl) Notify(e common.SensorComponentEvent) {
+}
+
+func (s *serviceImpl) Start() error {
+	log.Infof("Starting virtual machine component!")
+	ch2Central := make(chan *message.ExpiringMessage)
+	go func() {
+		defer func() {
+			s.stopper.Flow().ReportStopped()
+			close(ch2Central)
+		}()
+		for {
+			select {
+			case <-s.stopper.Flow().StopRequested():
+				return
+			case vm := <-s.fromDataSource:
+				s.toCentral <- message.New(&central.MsgFromSensor{
+					Msg: &central.MsgFromSensor_Event{
+						Event: &central.SensorEvent{
+							Id:     vm.Id,
+							Action: central.ResourceAction_SYNC_RESOURCE,
+							Resource: &central.SensorEvent_VirtualMachine{
+								VirtualMachine: vm,
+							},
+						},
+					},
+				})
+			}
+		}
+	}()
+	s.toCentral = ch2Central
+	return nil
+}
+
+func (s *serviceImpl) Stop() {
+	if !s.stopper.Client().Stopped().IsDone() {
+		defer utils.IgnoreError(s.stopper.Client().Stopped().Wait)
+	}
+	s.stopper.Client().Stop()
+}
+
+func (s *serviceImpl) Capabilities() []centralsensor.SensorCapability {
+	return []centralsensor.SensorCapability{}
```

Comment:
Use capabilities to let older Centrals know that they should ignore the virtual machine messages. In addition to feature flag, which is still useful to hotfix disable the feature.

Example `sensor/kubernetes/complianceoperator/update.go` line 120 as an example.

Other way around (Central -> Sensor capability) is not needed because Central does not send to Sensor.